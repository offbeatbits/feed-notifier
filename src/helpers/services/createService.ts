import { TwitterService, MastodonService } from "../../services";
import { serviceMappings } from "../../config";
import {
  ChannelName,
  ServiceClassSettings,
  MastodonSettings,
} from "../../types.d";

type ServiceInstance = TwitterService | MastodonService;

export const createService = (
  type: ChannelName,
  settings: ServiceClassSettings
) =>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (serviceMappings[type] as any)({
    // oof :(
    username: settings.username,
    host: (settings as MastodonSettings).host,
    settings: settings.client,
  }) as ServiceInstance;
