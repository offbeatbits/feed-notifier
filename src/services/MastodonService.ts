import generator, {
  MegalodonInterface,
} from 'megalodon';
import { OAuth2Service } from '../models';
import { MastodonServiceSettings } from '../types';

export class MastodonService extends OAuth2Service {
  username: string;

  host: string;

  private mastodonClient: MegalodonInterface;

  constructor({
    username,
    host,
    settings,
  }: MastodonServiceSettings) {
    super(settings);
    this.username = username;
    this.host = host;
    this.mastodonClient = generator(
      'mastodon',
      this.host,
      this.accessToken,
    );
  }

  async post(content: string) {
    const response = await this.mastodonClient.postStatus(content);
    return response.data.uri;
  }
}
