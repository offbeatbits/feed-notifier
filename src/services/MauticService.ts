/* eslint-disable class-methods-use-this */
import MauticConnector from 'node-mautic';
import { MauticServiceSettings } from '../types';

export class MauticService {
  url: string;

  username: string;

  password: string;

  private mauticClient: MauticConnector;

  constructor({
    url,
    username,
    password,
  }: MauticServiceSettings) {
    this.url = url;
    this.username = username;
    this.password = password;
    this.mauticClient = new MauticConnector({
      apiUrl: this.url,
      username: this.username,
      password: this.password,
      timeoutInSeconds: 20,
    });
  }

  async post(/* content: string */) {
    console.log(this.mauticClient); // eslint-disable-line
    return 'mautic service response';
  }
}
