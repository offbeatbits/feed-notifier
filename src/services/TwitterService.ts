/* eslint-disable class-methods-use-this */
import { TwitterClient } from "twitter-api-client";
import { OAuth2Service } from "../models";
import { TwitterServiceSettings } from "../types.d";

export class TwitterService extends OAuth2Service {
  username: string;

  private twitterClient: TwitterClient;

  constructor({ username, settings }: TwitterServiceSettings) {
    super(settings);
    this.username = username;
    this.twitterClient = new TwitterClient({
      apiKey: this.key,
      apiSecret: this.secret,
      accessToken: this.accessToken,
      accessTokenSecret: this.accessTokenSecret,
    });
  }

  async post(content: string) {
    const response = await this.twitterClient.tweets.statusesUpdate({
      status: content,
    });
    return `https://twitter.com/${this.username}/status/${response.id_str}`;
  }
}
