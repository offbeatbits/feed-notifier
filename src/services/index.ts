export { MastodonService } from './MastodonService';
export { TwitterService } from './TwitterService';
