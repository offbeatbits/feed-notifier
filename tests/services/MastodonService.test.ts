import { MastodonService } from "../../src/services/MastodonService";

jest.mock("megalodon", () =>
  jest.fn(() => ({
    postStatus: () =>
      Promise.resolve({
        data: {
          uri: "sample_uri_from_inline_megalodon_mock",
        },
      }),
  }))
);

describe("MastodonService", () => {
  const mastodonInstance = new MastodonService({
    enabled: true,
    username: "testUsername",
    host: "sample_host",
    settings: {
      key: "sample_key",
      secret: "sample_secret",
      accessToken: "sample_access_token",
    },
  });

  it("should be defined", () => {
    expect(MastodonService).toBeDefined();
  });

  it("should define post method", () => {
    expect(mastodonInstance.post).toBeDefined();
  });

  it("post method should return status URL", async () => {
    expect.assertions(1);
    const uri = await mastodonInstance.post("sample_status");
    expect(uri).toMatchSnapshot();
  });
});
