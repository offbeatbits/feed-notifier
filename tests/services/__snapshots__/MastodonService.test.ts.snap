// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MastodonService post method should return status URL 1`] = `"sample_uri_from_inline_megalodon_mock"`;
